// @chainx-v2/api/chainx/GateWayCommon.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

// eslint-disable-next-line header/header
// eslint-disable-next-line @typescript-eslint/no-unused-vars
class GateWayCommon {
  constructor(chainx) {
    this._api = void 0;
    this._api = chainx.api;
  }
  /**
   * query withdraw limit
   * */


  async withdrawalLimit(...args) {
    return new Promise(resolve => {
      // eslint-disable-next-line no-void
      void this._api.then(async api => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
        const withdrawLimits = await api.rpc.xgatewaycommon.withdrawalLimit(...args);
        resolve(withdrawLimits);
      });
    });
  }

  verifyWithdrawal(...args) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-call
    return new Promise(resolve => {
      // eslint-disable-next-line no-void
      void this._api.then(async api => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
        const veryfy = await api.rpc.xgatewaycommon.verifyWithdrawal(...args);
        resolve(veryfy);
      });
    });
  }

  trusteeMultisigs(...args) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-call
    return new Promise(resolve => {
      // eslint-disable-next-line no-void
      void this._api.then(async api => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
        const trusteeMultisigs = await api.rpc.xgatewaycommon.trusteeMultisigs(...args);
        resolve(trusteeMultisigs);
      });
    });
  }

  bitcoinTrusteeProperties(...args) {
    return new Promise(resolve => {
      // eslint-disable-next-line no-void
      void this._api.then(async api => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment
        const trusteeProperties = await api.rpc.bitcoinTrusteeProperties(...args);
        resolve(trusteeProperties);
      });
    });
  }

  bitcoinTrusteeSessionInfo(...args) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-call
    return new Promise(resolve => {
      // eslint-disable-next-line no-void
      void this._api.then(async api => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment
        const bitcoinTrusteeSessionInfo = await api.rpc.bitcoinTrusteeSessionInfo(...args);
        resolve(bitcoinTrusteeSessionInfo);
      });
    });
  }

  bitcoinGenerateTrusteeSessionInfo(...args) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-call
    return new Promise(resolve => {
      // eslint-disable-next-line no-void
      void this._api.then(async api => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment
        const bitcoinGenerateTrusteeSessionInfo = await api.rpc.bitcoinGenerateTrusteeSessionInfo(...args);
        resolve(bitcoinGenerateTrusteeSessionInfo);
      });
    });
  }

}

var _default = GateWayCommon;
exports.default = _default;
