// @chainx-v2/api/chainx/ChainX.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _rpcProvider = require("@chainx-v2/rpc-provider");

var _promise = _interopRequireDefault(require("../promise"));

// eslint-disable-next-line header/header
// import Account from '@chainx-v2/account';
class ChainX {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  // eslint-disable-next-line no-empty-pattern
  constructor(wsUrlOrProvider = 'ws://127.0.0.1:8087') {
    this._provider = void 0;
    this._api = void 0;
    this._provider = new _rpcProvider.WsProvider(wsUrlOrProvider);
  }

  getProvider() {
    return this._provider;
  } // eslint-disable-next-line no-empty-pattern


  setProvider(wsUrlOrProvider = 'ws://127.0.0.1:8087') {
    this._provider = new _rpcProvider.WsProvider(wsUrlOrProvider);
  }

  async ready() {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
    this._api = await _promise.default.create({
      provider: this._provider
    });
  }

  getApi() {
    return this._api;
  }

}

var _default = ChainX;
exports.default = _default;
